ADR 0070: Support for zstd:chunked image compression (#357)
konflux-ci/architecture: ADR/0070-dual-compression-for-container-builds.md

Both variant pushes must use `--format=oci`. Mixing OCI and Docker manifest formats within a single index causes "manifest invalid" error on Quay. If `COMPRESSION_FORMAT=dual` and the image format is not OCI, the task must detect this incompatible combination and error out early with a clear message.

`IMAGE_DIGEST` points to the per-arch index, which contains both variants. Because Tekton Chains does not currently support recursive signing of child manifests within an index ([tektoncd/chains#1070](https://github.com/tektoncd/chains/issues/1070)), the build task explicitly exports the digests for both the gzip and zstd child manifests in its `IMAGES` result. This allows Chains to generate provenance for each manifest directly from the trusted build task, closing potential trust gaps. Downstream tasks, SBOM generation, and signing consume `IMAGE_DIGEST` as the primary artifact. The `io.github.containers.compression.zstd` annotation is set on the zstd manifest descriptor within the index, which is how Podman and CRI-O [identify and prefer](https://github.com/containers/image/blob/main/internal/manifest/oci_index.go) the zstd variant.

> In `gzip` or `zstd-chunked` mode, the build task pushes a single manifest as it does today. `IMAGE_DIGEST` holds the primary digest (gzip in `gzip` mode, zstd in `zstd-chunked` mode).

Author: It is existing tekton result convention used by both tasks today

The `build-image-index` task requires no new parameters. Each entry in the existing `IMAGES` array is now a per-arch index (in `dual` mode) instead of a single manifest. The task's existing `--all` flag on `buildah manifest add` is expected to extract child manifests from per-arch indexes and flatten them into the final multi-arch index, preserving the gzip-first invariant or explicit flattening logic will be added in the task if needed.

SBOM generation requires an update; the SBOM must attach to both the gzip and zstd child manifest digests. The per-arch index is an intermediate build artifact that gets flattened by `build-image-index`, so it is not a valid long-term attachment point. Both compression variants receive the same SBOM since they share identical content, only differing in layer compression.

> 

Author: Ack, implementation will include post assembly validation, ADR intentionally leaves room for this

## Consequences

Pros:

* Faster pulls for modern clients. Users with Podman and CRI-O will automatically benefit from zstd:chunked without changing their pipeline configuration. Docker 23.0+ clients gain tolerance for the zstd layer, but will continue to receive the gzip variant via first-match selection.
* Full backward compatibility. Old Docker clients continue to work because gzip is listed first in the image index.
* Provenance and signing. Because the build task explicitly exports the child manifests in `IMAGES`, Tekton Chains can generate provenance for each manifest directly from the trusted build task, avoiding trust gaps where manifests are only discovered by downstream tasks.

Cons:

* The final image index contains 2N manifests instead of N (one per architecture per compression format). Exposing all of these in the `IMAGES` result increases the Tekton result size, pushing closer to the 4KB limit for large matrix builds (e.g., builds exceeding 4-5 architectures may hit this limit).
* Registry storage per image roughly doubles. Each layer is stored in both gzip and zstd formats. Users who don't want this can keep `COMPRESSION_FORMAT=gzip`.

> * Builds take longer in dual mode. The zstd push recompresses all layers, adding time proportional to image size.

Reviewer: By default, we only add one more layer to the images. If we are recompressing all layers, will we lose this connection to the parent image?

I imagine so if the parent image doesn't have zstd:chunked, but if the parent image has that compression available, will this still be a consequence (longer builds due to recompressing) or will we be able to take advantage of the parent images' builds?

Author: I think yes, only when the parent image doesn't already have zstd:chunked layers, in that case, the parent layers need to be recompressed, when zstd:chunked layers are already available, they can be reused directly, so we still get the layer sharing benefits and only new layers need compression.

Reviewer: Is there any additional logic needed to reuse the zstd:chunked compression from the parent image or that "just works?"

Author: After looking into it more, the parent's zstd layers actually cannot be reused, Buildah stores all layers uncompressed locally and compresses them fresh at push time, when we specify --compression-format zstd:chunked, it forces recompression of all layers regardless, there's no way to check the registry for existing zstd blobs and skip the work, so the "longer builds" consequence holds either way

`IMAGE_DIGEST` points to the per-arch index, which contains both variants. Because Tekton Chains does not currently support recursive signing of child manifests within an index ([tektoncd/chains#1070](https://github.com/tektoncd/chains/issues/1070)), the build task explicitly exports the digests for both the gzip and zstd child manifests in its `IMAGES` result. This allows Chains to generate provenance for each manifest directly from the trusted build task, closing potential trust gaps. Downstream tasks, SBOM generation, and signing consume `IMAGE_DIGEST` as the primary artifact. The `io.github.containers.compression.zstd` annotation is set on the zstd manifest descriptor within the index, which is how Podman and CRI-O [identify and prefer](https://github.com/containers/image/blob/main/internal/manifest/oci_index.go) the zstd variant.

In `gzip` or `zstd-chunked` mode, the build task pushes a single manifest as it does today. `IMAGE_DIGEST` holds the primary digest (gzip in `gzip` mode, zstd in `zstd-chunked` mode).

> The `build-image-index` task requires no new parameters. Each entry in the existing `IMAGES` array is now a per-arch index (in `dual` mode) instead of a single manifest. The task's existing `--all` flag on `buildah manifest add` extracts child manifests from per-arch indexes and flattens them into the final multi-arch index. The `--all` preserves both the gzip-first ordering and the `io.github.containers.compression.zstd` annotation on the zstd child manifest descriptors. If [ADR 0069](https://github.com/konflux-ci/architecture/pull/359) proposes a platform sort in `build-image-index` for reproducibility, it must either include compression type in the sort key or use a stable sort to preserve the gzip-first invariant.

Reviewer 2: **[nit]** ADR 0069 is merged now, so this could be declarative. The link could also point to the file instead of the closed PR.

```suggestion
The `build-image-index` task requires no new parameters. Each entry in the existing `IMAGES` array is now a per-arch index (in `dual` mode) instead of a single manifest. The task's existing `--all` flag on `buildah manifest add` extracts child manifests from per-arch indexes and flattens them into the final multi-arch index. The `--all` preserves both the gzip-first ordering and the `io.github.containers.compression.zstd` annotation on the zstd child manifest descriptors. [ADR 0069](0069-reproducible-container-builds.md) proposes a platform sort in `build-image-index` for reproducibility. That sort must either include compression type in the sort key or use a stable sort to preserve the gzip-first invariant.
```

This was originally proposed in [build-definitions#1264](https://github.com/konflux-ci/build-definitions/issues/1264). It was also demanded in [build-definitions#3188](https://github.com/konflux-ci/build-definitions/issues/3188) for large CUDA images, and [release-service-catalog#634](https://github.com/konflux-ci/release-service-catalog/pull/634) already noted the need for zstd layer handling in Pyxis.

## Decision

Add a dual-compression setting to the build pipeline. In `dual` mode, the `buildah` task pushes both gzip and zstd:chunked variants, bundles them into a per-arch OCI image index, and exposes the index digest as `IMAGE_DIGEST`. No new Tekton results are needed. It also risks breaking release pipeline tasks that assume one manifest per architecture, so those need deduplication fixes before the default can change.

### Implementation details

The `buildah` task will get 2 new params:
* `COMPRESSION_FORMAT` — `gzip` (default), `zstd-chunked`, or `dual`
* `FORCE_COMPRESSION` — when `true`, recompresses all layers, including the base image layers

> In `dual` mode, the push step creates a **per-arch index** containing both compression variants:

Reviewer 3: Were you able to test what effect this has on the post-build tasks that process the output image? Are they able to process a gzip+zstd index instead of a single manifest?

Author: Not yet, this is on the testing plan, downstream tasks handle indexes in multi arch builds, but i'll check for single arch case and report ASAP

Reviewer: If we cannot resolve this in the ADR then I think it would be fine to call it out as an unknown. Would it fit as a potential consequence or would it be better elsewhere?
